### ISSUE TITLE
OpenID Connect

### ISSUE THREAD

**Issue by Mirtaaa** [zammad/zammad#4997]
#### Introduction

> OpenID is an easy and safe way for people to reuse an existing account and user profile from an identity provider, for example Apple, Google, or Microsoft to sign-in to any OpenID-enabled applications and websites without creating a new registration and password. You choose the provider, such as Google and enter your Gmail address and password to sign-in.

For further information see https://openid.net/developers/discover-openid-and-openid-connect/.

#### Use Cases

* I want to use an existing OpenID Connect client (e.g. from a Keycloak server) to login into Zammad.

#### Intended Outcome

The integration of OpenID Connect enables users to connect Zammand to an OP (OpenID provider) in an easy, reliable, and future-proof way.

#### Acceptance Criteria

* Configure one OP in Zammad that needs to support [OpenID Connect Discovery](https://openid.net/specs/openid-connect-discovery-1_0.html)
  * No manual configuration of the OP for now
* Possibility to change the default UserID claim (defaults to `sub`)
* Possibility to configure the scopes (defaults to `openid`, `email` and `profile`)
* Logout in Zammad is logging out the related user in the OP
* Logout in the OP is logging out the related user in Zammad

#### Mockups

<img width="464" alt="image" src="https://github.com/user-attachments/assets/e5b39f0f-8630-4846-aa40-25115237d609">

![image](https://github.com/user-attachments/assets/c13a439c-224b-4657-9b5c-d34276b2d7c7)

**Comment by fliebe92** [comment]
### Use Case

* Use OIDC in Zammad for authentication (as e.g. SAML, Google, ...)

### Acceptance Criteria

* Single vs multiple OpenID provider configurations
* Support for simplifying configuration using OpenID Connect Discovery: https://openid.net/specs/openid-connect-discovery-1_0.html
* Support for overriding the default UserID claim (defaults to `sub`)
* Support for configuring scopes (defaults to `openid`, `email` and `profile`)

**Comment by benclark158** [comment]
Hi, I'm just wondering if OIDC is likely to be implemented into Zammad soon? I am hoping to deploy Zammad into my stack in the new year, but need connect it with our OIDC provider which doesn't support SAML. Thanks in advance.

**Comment by fliebe92** [comment]
Hi @benclark158. OIDC will be part of the next release of Zammad. I'm sorry, but you still have to wait for it.

**Comment by benclark158** [comment]
Hi @fliebe92. Thank you for getting it implemented! No worries on waiting for it :)

I've been trying out the `develop` branch via docker today but I have run into an issue. I haven't been able to set the client secret needed for OIDC to get the token. Is there a way to set it at the moment?

**Comment by fliebe92** [comment]
> I've been trying out the develop branch via docker today but I have run into an issue. I haven't been able to set the client secret needed for OIDC to get the token. Is there a way to set it at the moment?

Hi @benclark158. Currently, we're supporting OIDC with the discovery feature in place. No chance to enter the client secret manually. May I ask what OP are you using?

You can also check the related documentation for OIDC in Zammad: https://admin-docs.zammad.org/en/pre-release/settings/security/third-party/openid-connect.html

**Comment by benclark158** [comment]
Hi @fliebe92. I looked into this as we use a custom OP.

I tried it with keycloak today and it successfully worked. I am guessing keycloak supports part of the OIDC spec that our provider doesn't. I am going to look more into the spec and hopefully find the issue in our OP. Thank you for your help!

**Comment by mweinelt** [comment]
I couldn't find this out based on the code, but the settings define a public OIDC client, so I wonder if it supports [PKCE](https://www.rfc-editor.org/rfc/rfc7636). Our IdP ([kanidm](https://github.com/kanidm/kanidm)) strictly [enforces PKCE on public clients](https://kanidm.github.io/kanidm/master/integrations/oauth2.html#public-client-configuration).

**Comment by fliebe92** [comment]
> I couldn't find this out based on the code, but the settings define a public OIDC client, so I wonder if it supports [PKCE](https://www.rfc-editor.org/rfc/rfc7636). Our IdP (kanidm) strictly enforces PKCE on public clients.

Hi @mweinelt. No, PKCE is currently not supported.

**Comment by jinnatar** [comment]
FYI, for public clients PKCE is required to mitigate real world attacks.

TL;DR; PKCE is a very good idea both for public and confidential setups but absolutely necessary for public. If it seems annoying to implement, the other option is implementing confidential clients to have some base level of security.

**Comment by fliebe92** [comment]
> FYI, for public clients PKCE is required to mitigate real world attacks.
> 
> TL;DR; PKCE is a very good idea both for public and confidential setups but absolutely necessary for public. If it seems annoying to implement, the other option is implementing confidential clients to have some base level of security.

Hi @jinnatar. This is just the first iteration about OpenID Connect in Zammad, to make it work in a simple way and just to get started with it.

**Comment by YetAnotherGerrit** [comment]
PKCE is mainly to protect the process in two scenarios:
- When the client secret is exposable, which is mainly relevant in local apps, like mobile apps or single page applications.
- When a man-in-the-middle attack is applicable.

From my understand, we have a very low risk here since:
- The client secret is stored in the backend.
- http is hopefully not used anywhere in the process.

Correct?

**Comment by jinnatar** [comment]
> PKCE is mainly to protect the process in two scenarios:
> - When the client secret is exposable, which is mainly relevant in local apps, like mobile apps or single page applications.
> - When a man-in-the-middle attack is applicable.
> 
> From my understand, we have a very low risk here since:
> - The client secret is stored in the backend.

In public client mode there is no client secret, implementing a client secret stored in the backend requires supporting confidential clients which indeed would partially mitigate the attack scenarios.

> - http is hopefully not used anywhere in the process.
> 
> Correct?

PKCE specifically addresses the parts of the exchange where TLS cannot protect the user, for example local compromise. The RFC linked above has a decent explanation and graph of the exchange and attack vectors. Real world token stealing attacks use this vector.

**Comment by dominikklein** [comment]
We will implement PKCE support with https://github.com/zammad/zammad/issues/5485

**Comment by leeoocca** [comment]
Thanks for the work on this feature!

> OIDC will be part of the next release of Zammad. I'm sorry, but you still have to wait for it.

When is the next release planned for? I couldn't estimate it from the last releases https://zammad.com/en/releases as it doesn't seem to follow a regular interval. From what I can see in the milestones https://github.com/zammad/zammad/milestones both 6.4 and 6.5 seem to be ready.

Thanks again for all your work 🙏🏼

**Comment by fliebe92** [comment]
Hi @leeoocca. There is currently no date I can tell you. Please be patient. :)

**Comment by leeoocca** [comment]
Hi @fliebe92, sorry for the impatience, it would allow us to finally start using your amazing product.

Is there a way to already deploy and try this, maybe a canary release? I see the `6.5.0-alpha` was pushed recently to Docker Hub: https://hub.docker.com/layers/zammad/zammad-docker-compose/6.5.0-alpha/images/sha256-7599a59f2820806784056ae4f99a3fe26b00805c44fcc6f081a7e4375c7661a4

**Comment by fliebe92** [comment]
Hi @leeoocca. OpenID Connect is only merged to branches that are not yet ready for production usage. Therefore, IMO, you still have to wait. :/

**Comment by pirhoo** [comment]
> Hi [@fliebe92](https://github.com/fliebe92). I looked into this as we use a custom OP.
> 
> I tried it with keycloak today and it successfully worked. I am guessing keycloak supports part of the OIDC spec that our provider doesn't. I am going to look more into the spec and hopefully find the issue in our OP. Thank you for your help!

Hey, did you find what are the differences with Keycloak? We are trying to set this up as well (with [Doorkeeper](https://github.com/doorkeeper-gem/doorkeeper-openid_connect)) but the missing secret is a blocker.

**Comment by herbetom** [comment]
I'm mentioning it because it went completely unnoticed by me: About a month ago, 6.5 was released with OpenID Connect support: https://zammad.com/de/releases/6-5#3-nahtloses-login-mit-openid-connect

Haven't tested it yet though.

Just in case others also didn't notice it. :)

**Comment by mweinelt** [comment]
I took the time to test this and it works great!<del> and the only issue I found was that we need to be able to merge OIDC users with existing users. Deleting existing users is not great, since that deletes all tickets they've created and unassigns all tickets they're assigned to.</del>

**Comment by herbetom** [comment]
I guess you could create a temp user, use for example a scheduler to move all tickets over, delete the old user, let the user sign in via the new method (that's probably the most annoying part) and use another scheduler to move all tickets to the new user.

But quite a manual and involved procces and merging would be certainly better. 👍

**Comment by mweinelt** [comment]
Thanks for poking me in private. Enabling account linking on the security tab makes it work. Awesome!

https://admin-docs.zammad.org/de/latest/settings/security/third-party.html#automatic-account-link-on-initial-logon
